Export pid gains as reference

kumar-sanjeeev opened the pull request:
This PR includes the following changes:

- exported p, i, d gains as reference.
- Updated test cases.

kumar-sanjeeev committed:
Add bool param to export gains as reference

kumar-sanjeeev committed:
Add gains to export as references based on export_gain_reference param value

kumar-sanjeeev committed:
Add export_gain_references param in yaml

christophfroehlich reviewed (commented):
In principle, this looks fine for me.

TheDevMystic commented on pid_controller/include/pid_controller/pid_controller.hpp:
```diff
@@ -76,6 +76,9 @@ class PidController : public controller_interface::ChainableControllerInterface
   std::vector<double> measured_state_values_;
 
+  inline static const std::vector<std::string> GAIN_INTERFACES = {"p", "i", "d"};
+  inline static const std::vector<size_t> GAIN_TYPES_INDEX = {0, 1, 2};
```
A quick question about convention of ROS Controls, do we follow,

`[Storage Specificier] [Modifier] [Type] [Variable Name]`
Like: static constexpr bool is_ros = true;

Or
`[Modifier] [Storage Specificier] [Type] [Variable Name]`

Like: constexpr static bool is_ros = true;

If we follow the above convention, then I think you should change the order of specifier. 

Am I missing something, or is the code convention different. 

I assumed the first convention because it is used pre-dominantely in other codebase.

Thanks!

TheDevMystic commented:
I also think implementation is correct @kumar-sanjeeev, 
But if implementation is correct then, why are we failing tests?

Overall, I think some tests need to be updated, but this is just my assumption.

kumar-sanjeeev committed:
Switch modifier and storage specifier position to maintain codebase consistency

kumar-sanjeeev committed:
Do not skip gains by setting them NaN

kumar-sanjeeev committed:
Update test cases to accomodate for both values of exported_gain_references param

kumar-sanjeeev committed:
Shift export_gain_references variable to fix abi check

kumar-sanjeeev committed:
Add gain references inside new key to pass abi check

kumar-sanjeeev committed:
Fix variable name as per gtest convention

thedevmystic commented on pid_controller/src/pid_controller.cpp:
```diff
@@ -495,6 +517,43 @@ controller_interface::return_type PidController::update_and_write_commands(
+      auto current_pid_gains = pids_[i]->get_gains();
+      for (size_t j = 0; j < GAIN_INTERFACES.size(); ++j)
+      {
+        const size_t buffer_index = gains_start_index + i + j * dof_;
```
```suggestion
        const size_t buffer_index = gains_start_index + (j * dof_) + i;
```

As the standard formula for index calculation is,
`index = base_offset + (row_index * row_size) + column_index`.

While the previous one is perfectly fine, this improves general readability and precedence readability.

kumar-sanjeeev committed:
Add code review suggestions

kumar-sanjeeev committed:
Update docs with gain reference export possibility

christophfroehlich commented on pid_controller/src/pid_controller.cpp:
```diff
@@ -495,6 +517,43 @@ controller_interface::return_type PidController::update_and_write_commands(
+              current_pid_gains.d_gain_ = new_gain_value;
+              pids_[i]->set_gains(current_pid_gains);
+              break;
+          }
```
This is part is not covered by tests. Can you please think of a testcase in doing so? Friending the class and checking pids_ variable might be possible.

kumar-sanjeeev replied:
Hi @christophfroehlich,

I was on vacation, will into comments and get back to you

christophfroehlich reviewed (changes requested):
Another quick look and question

kumar-sanjeeev commented:
> @kumar-sanjeeev any updates on this?

@christophfroehlich apologies, this PR slipped through my mind. I'll work on it in coming days and get it closed.

Commit:
Remove 'export_params' parameter and modified logic accordingly

kumar-sanjeeev committed:
Update docs

kumar-sanjeeev commented:
@christophfroehlich,

Sorry for the delay. Based on your feedback, I have removed the added parameter and modified the logic. Also added a corresponding test case for the mentioned code.

kumar-sanjeeev commented:
hi @christophfroehlich,

when you will get time, can you please provide feedback on last committed changes.

christophfroehlich commented on pid_controller/src/pid_controller.cpp:
```diff
@@ -533,6 +548,40 @@ controller_interface::return_type PidController::update_and_write_commands(
+      if (std::isfinite(new_gain_value))
+      {
+        const size_t gain_type = GAIN_TYPES_INDEX[j];
+        switch (gain_type)
```
this could be done by changing GAIN_INTERFACES to a `std::vector<char>`, which then could be used in the switch-case directly without the need of GAIN_TYPES_INDEX?

christophfroehlich reviewed (commented):
Please never do force pushes if PRs are already under review. It breaks the github diff view and just takes more time for review.

Overall LGTM, one suggestion for further cleanup. what do you think?

christophfroehlich committed:
Merge branch 'master' into feat-export-pid-gains-as-reference

christophfroehlich committed:
Refactor GAIN_TYPE indexing
